sklearn/linear_model/cd_fast.pyx

```diff
@@ -143,15 +153,118 @@ cdef extern from "cblas.h":
     void sscal "cblas_sscal"(int N, float alpha, float *X, int incX) nogil
 
 
+# Function to compute the duality gap
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef inline floating enet_duality_gap(
+            # Data
+            unsigned int n_samples,
+            unsigned int n_features,
+            unsigned int n_tasks,
+            floating* X_data,
+            floating* y_data,
+            floating* R_data,
+            floating* w_data,
+            # Variables intended to be modified
+            floating* XtA_data,
+            floating* dual_scaling,
+            # Parameters
+            floating alpha,
+            floating beta,
+            bint positive,
+            # active set for improved performance
+            int[:] disabled,
+            unsigned int n_disabled = 0
+            ) nogil:
+
+    cdef floating R_norm2
+    cdef floating w_norm2
+    cdef floating y_norm2
+    cdef floating l1_norm
+    cdef floating cst
+    cdef floating gap
+    cdef floating yTR
+    cdef floating dual_norm_XtA
+    cdef unsigned int i
+
+    # fused types version of BLAS functions
+    cdef DOT dot
+    cdef ASUM asum
+
+    if floating is float:
+        dot = sdot
+        asum = sasum
+    else:
+        dot = ddot
+        asum = dasum
+
+    # XtA = np.dot(X.T, R) - beta * w
+    for i in range(n_features):
+        if (n_disabled == 0 or disabled[i] == 0):
+            XtA_data[i] = dot(n_samples, &X_data[i * n_samples],
+                              1, R_data, 1) - beta * w_data[i]
+        else:
+            # We only need the infinity norm of XtA and by KKT, we know that
+            # in the present case, XtA[i] will not reach the maximum
+            XtA_data[i] = 0
+
+    yTR = dot(n_samples, y_data, 1, R_data, 1)
+
+    # R_norm2 = np.dot(R, R)
+    R_norm2 = dot(n_samples, R_data, 1, R_data, 1)
+
+    if alpha == 0:
+        dual_scaling[0] = 1
+    else:
+        if positive:
+            dual_norm_XtA = max(n_features, XtA_data)
+        else:
+            dual_norm_XtA = abs_max(n_features, XtA_data)
+
+        if dual_norm_XtA <= 0:
+            if R_norm2 == 0:
+                dual_scaling[0] = 1. / alpha
+            else:
+                dual_scaling[0] = yTR / (R_norm2 * alpha)
+        elif positive:
+            dual_scaling[0] = fmin(yTR / (alpha * R_norm2),
+                                   1. / dual_norm_XtA)
+        else:
+            dual_scaling[0] = fmin(fmax(yTR / (alpha * R_norm2),
+                                        -1. / dual_norm_XtA),
+                                   1. / dual_norm_XtA)
+
+    dual_scaling[0] = 1. / dual_scaling[0]
+
+    # w_norm2 = np.dot(w, w)
+    w_norm2 = dot(n_features, w_data, 1, w_data, 1)
+
+    l1_norm = asum(n_features, w_data, 1)
+
+    # np.dot(R.T, y)
+    cst = alpha / dual_scaling[0]
+    gap = (alpha * l1_norm - cst * yTR +
+           0.5 * (1. + cst ** 2) * (R_norm2 + beta * w_norm2))
+
```

Comment:
If I use the dual definition proposed in from the [test](https://github.com/scikit-learn/scikit-learn/pull/7853/files#diff-ff4438be9ee77932848f5abfcec060efR839), I found : 
```python
gap = (alpha * l1_norm - cst * yTR +
       0.5 * (1. + cst ** 2) * R_norm2 + 
       0.5 * (1. + alpha ** 2) * beta * w_norm2))
```
Am I right ?